crates/sui-indexer-alt-framework/src/pipeline/concurrent/pruner.rs

```diff
@@ -703,4 +826,103 @@ mod tests {
+            checkpoint_hi_inclusive: 4,
+            tx_hi: 8,
+            timestamp_ms_hi_inclusive: timestamp,
+            reader_lo: 4,        // Allow pruning up to checkpoint 4 (exclusive)
+            pruner_timestamp: 0, // Past timestamp so delay doesn't block
+            pruner_hi: 1,
+        };
+
+        // Configure failing behavior: range [1,2) should fail once before succeeding
+        let mut remaining_failures = HashMap::new();
+        remaining_failures.insert((1, 2), 1);
+        let store = MockStore {
+            watermarks: Arc::new(Mutex::new(watermark)),
+            data: Arc::new(Mutex::new(test_data.clone())),
+            remaining_failures: Arc::new(Mutex::new(remaining_failures)),
+        };
+
+        // Start the pruner
+        let store_clone = store.clone();
+        let cancel_clone = cancel.clone();
+        let pruner_handle = tokio::spawn(async move {
+            pruner(
+                handler,
+                Some(pruner_config),
+                store_clone,
+                metrics,
+                cancel_clone,
+            )
+            .await
+        });
+
+        // Wait for first pruning cycle - ranges [2,3) and [3,4) should succeed, [1,2) should fail
+        tokio::time::sleep(Duration::from_millis(500)).await;
+        {
+            let data = store.data.lock().unwrap();
+            let watermarks = store.watermarks.lock().unwrap();
+
+            // Verify watermark doesn't advance past the failed range [1,2)
+            assert_eq!(
+                watermarks.pruner_hi, 1,
+                "Pruner watermark should remain at 1 because range [1,2) failed"
+            );
+            assert!(data.contains_key(&1), "Checkpoint 1 should be preserved");
+            assert!(!data.contains_key(&2), "Checkpoint 2 should be pruned");
+            assert!(!data.contains_key(&3), "Checkpoint 3 should be pruned");
+            assert!(data.contains_key(&4), "Checkpoint 4 should be preserved");
+        }
+
+        // Wait for second pruning cycle - range [1,2) should succeed on retry
+        tokio::time::sleep(Duration::from_millis(3000)).await;
+        {
+            let data = store.data.lock().unwrap();
+            let watermarks = store.watermarks.lock().unwrap();
+
+            // Verify watermark advances after all ranges complete successfully
+            assert_eq!(
+                watermarks.pruner_hi, 4,
+                "Pruner watermark should advance to 4 after all ranges complete"
+            );
+            assert!(!data.contains_key(&1), "Checkpoint 1 should be pruned");
+            assert!(!data.contains_key(&2), "Checkpoint 2 should be pruned");
+            assert!(!data.contains_key(&3), "Checkpoint 3 should be pruned");
+            assert!(data.contains_key(&4), "Checkpoint 4 should be preserved");
+        }
+
+        // Clean up
+        cancel.cancel();
+        let _ = tokio::time::timeout(Duration::from_millis(1000), pruner_handle).await;
```

Comment:
nit: do we need this step? It seems to me that either we should care that the handle winds down at all costs (in which case we shouldn't have the timeout), or we should be okay with it (in which case, we probably don't need to wait for the handle to shut down at all).